<a href="https://colab.research.google.com/github/HarshESC/Deep-Learning/blob/main/Deep_Learning_Fashion_Minst.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:

import warnings

warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD



In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()
input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype("float32")
X_test = X_test.astype("float32")
X_train /= 255
X_test /= 255

y_train = to_categorical(y_train, nb_classes)
y_test = to_categorical(y_test, nb_classes)


11501568/11490434 [==============================] - 0s 0us/step


In [3]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))


In [4]:
model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])


In [5]:
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 4s 4ms/step - loss: 1.3185 - accuracy: 0.6564
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5093 - accuracy: 0.8690
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3894 - accuracy: 0.8933
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3421 - accuracy: 0.9035
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3144 - accuracy: 0.9109
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2945 - accuracy: 0.9166
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2787 - accuracy: 0.9209
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2654 - accuracy: 0.9244
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2541 - accuracy: 0.9280
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2439 - accuracy: 0.9309

In [6]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.1763189136981964
Test accuracy: 0.9506999850273132


In [7]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.9751 - accuracy: 0.3263
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.9627 - accuracy: 0.7364
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4909 - accuracy: 0.8581
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3856 - accuracy: 0.8878
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3326 - accuracy: 0.9030
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2978 - accuracy: 0.9123
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2722 - accuracy: 0.9204
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2506 - accuracy: 0.9269
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2329 - accuracy: 0.9327
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2175 - accuracy: 0.9371

In [8]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.1418158859014511
Test accuracy: 0.9585999846458435


In [9]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 3s 5ms/step - loss: 2.1151 - accuracy: 0.2144
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 0.9204 - accuracy: 0.7230
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.4385 - accuracy: 0.8749
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 0.3185 - accuracy: 0.9081
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2552 - accuracy: 0.9280
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2142 - accuracy: 0.9387
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 0.1831 - accuracy: 0.9472
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 0.1607 - accuracy: 0.9537
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 0.1431 - accuracy: 0.9589
Epoch 10/20
469/469 [==============================] - 2s 5ms/step - loss: 0.1288 - accuracy: 0.9632

In [10]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.10351189970970154
Test accuracy: 0.9703999757766724


In [11]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(1024, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 4s 8ms/step - loss: 1.6487 - accuracy: 0.4416
Epoch 2/20
469/469 [==============================] - 4s 8ms/step - loss: 0.4682 - accuracy: 0.8644
Epoch 3/20
469/469 [==============================] - 4s 8ms/step - loss: 0.2855 - accuracy: 0.9185
Epoch 4/20
469/469 [==============================] - 4s 8ms/step - loss: 0.2153 - accuracy: 0.9379
Epoch 5/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1714 - accuracy: 0.9503
Epoch 6/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1446 - accuracy: 0.9585
Epoch 7/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1227 - accuracy: 0.9653
Epoch 8/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1044 - accuracy: 0.9697
Epoch 9/20
469/469 [==============================] - 4s 8ms/step - loss: 0.0900 - accuracy: 0.9740
Epoch 10/20
469/469 [==============================] - 4s 8ms/step - loss: 0.0785 - accuracy: 0.9777

In [12]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.0821097269654274
Test accuracy: 0.9761999845504761


The most accurate model came from using 9 layers. The speed of the model was much slower. The run time of a 7 layer model was noticeably faster but performed with a worse accuracy.



# Tanh activation and SGD loss


In [13]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(10, activation="softmax"))


In [14]:
model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])


In [15]:
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.0531 - accuracy: 0.7479
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5291 - accuracy: 0.8702
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4246 - accuracy: 0.8887
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3763 - accuracy: 0.8974
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3472 - accuracy: 0.9034
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3267 - accuracy: 0.9081
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3110 - accuracy: 0.9116
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2981 - accuracy: 0.9151
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2872 - accuracy: 0.9177
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2776 - accuracy: 0.9206

In [16]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.21235273778438568
Test accuracy: 0.9384999871253967


In [17]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.3666 - accuracy: 0.6505
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.7486 - accuracy: 0.8396
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5529 - accuracy: 0.8728
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4585 - accuracy: 0.8893
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4011 - accuracy: 0.9003
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3615 - accuracy: 0.9077
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3315 - accuracy: 0.9140
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3078 - accuracy: 0.9191
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2881 - accuracy: 0.9239
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2711 - accuracy: 0.9280

In [18]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.17822496592998505
Test accuracy: 0.9502000212669373


In [19]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(256, activation="tanh"))
model.add(Dense(128, activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 3s 5ms/step - loss: 1.0918 - accuracy: 0.7685
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 0.5886 - accuracy: 0.8812
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.4305 - accuracy: 0.9020
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 0.3530 - accuracy: 0.9148
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.3050 - accuracy: 0.9238
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2705 - accuracy: 0.9308
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2439 - accuracy: 0.9371
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2223 - accuracy: 0.9421
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2036 - accuracy: 0.9462
Epoch 10/20
469/469 [==============================] - 2s 5ms/step - loss: 0.1877 - accuracy: 0.9501

In [20]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.11343056708574295
Test accuracy: 0.9675999879837036


In [21]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(1024, activation="tanh"))
model.add(Dense(512, activation="tanh"))
model.add(Dense(256, activation="tanh"))
model.add(Dense(128, activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 4s 8ms/step - loss: 0.9598 - accuracy: 0.8180
Epoch 2/20
469/469 [==============================] - 4s 8ms/step - loss: 0.4948 - accuracy: 0.8995
Epoch 3/20
469/469 [==============================] - 4s 8ms/step - loss: 0.3602 - accuracy: 0.9180
Epoch 4/20
469/469 [==============================] - 4s 8ms/step - loss: 0.2920 - accuracy: 0.9300
Epoch 5/20
469/469 [==============================] - 4s 8ms/step - loss: 0.2469 - accuracy: 0.9401
Epoch 6/20
469/469 [==============================] - 4s 8ms/step - loss: 0.2140 - accuracy: 0.9477
Epoch 7/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1881 - accuracy: 0.9536
Epoch 8/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1670 - accuracy: 0.9588
Epoch 9/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1500 - accuracy: 0.9627
Epoch 10/20
469/469 [==============================] - 4s 8ms/step - loss: 0.1360 - accuracy: 0.9653

In [22]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.09278957545757294
Test accuracy: 0.9739999771118164


The model with 9 layers had the best test accuracy, however it was noticeably slower. Additionally, the accuracy was only better by around 0.01. Using tanh activation provided very similar results and the accuracy was just barely lower than when using ReLU.



# Sigmoid activation and SGD loss¶


In [23]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2775 - accuracy: 0.1776
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2277 - accuracy: 0.3408
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 2.1674 - accuracy: 0.4741
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 2.0816 - accuracy: 0.5420
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 1.9591 - accuracy: 0.5692
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 1.7976 - accuracy: 0.6012
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 1.6120 - accuracy: 0.6327
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 1.4292 - accuracy: 0.6633
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 1.2697 - accuracy: 0.6965
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 1.1393 - accuracy: 0.7252

In [24]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.5689396262168884
Test accuracy: 0.8528000116348267


In [25]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3196 - accuracy: 0.1071
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3007 - accuracy: 0.1124
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3004 - accuracy: 0.1124
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3002 - accuracy: 0.1124
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3000 - accuracy: 0.1124
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2998 - accuracy: 0.1124
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2996 - accuracy: 0.1124
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2994 - accuracy: 0.1124
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2992 - accuracy: 0.1124
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 2.2989 - accuracy: 0.1124

In [26]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 2.2951602935791016
Test accuracy: 0.11349999904632568


In [27]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(256, activation="sigmoid"))
model.add(Dense(128, activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 3s 5ms/step - loss: 2.3298 - accuracy: 0.0962
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3015 - accuracy: 0.1124
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 10/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124

In [28]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 2.3010759353637695
Test accuracy: 0.11349999904632568


In [29]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(1024, activation="sigmoid"))
model.add(Dense(512, activation="sigmoid"))
model.add(Dense(256, activation="sigmoid"))
model.add(Dense(128, activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)


Epoch 1/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3357 - accuracy: 0.1100
Epoch 2/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3015 - accuracy: 0.1124
Epoch 3/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 4/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 5/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 6/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 7/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3014 - accuracy: 0.1124
Epoch 8/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 9/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 10/20
469/469 [==============================] - 4s 8ms/step - loss: 2.3013 - accuracy: 0.1124

In [30]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 2.301055431365967
Test accuracy: 0.11349999904632568


The test accuracy for the sigmoid tests are all pretty low. It appears using sigmoid activation is not a good idea for this model.



#ReLU activation and SGD loss Changing Batch Sizes¶


In [31]:
# Using 64 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 5s 4ms/step - loss: 1.1610 - accuracy: 0.6131
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.3354 - accuracy: 0.9029
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2329 - accuracy: 0.9331
Epoch 4/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1788 - accuracy: 0.9484
Epoch 5/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1459 - accuracy: 0.9571
Epoch 6/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1231 - accuracy: 0.9632
Epoch 7/20
938/938 [==============================] - 4s 5ms/step - loss: 0.1056 - accuracy: 0.9686
Epoch 8/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0907 - accuracy: 0.9732
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0790 - accuracy: 0.9765
Epoch 10/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0687 - accuracy: 0.9797

In [32]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.09467988461256027
Test accuracy: 0.9768999814987183


In [33]:
# Using 32 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=32, epochs=20, verbose=1)


Epoch 1/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.7273 - accuracy: 0.7735
Epoch 2/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.2070 - accuracy: 0.9399
Epoch 3/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1392 - accuracy: 0.9591
Epoch 4/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1054 - accuracy: 0.9686
Epoch 5/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0821 - accuracy: 0.9761
Epoch 6/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0672 - accuracy: 0.9802
Epoch 7/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0534 - accuracy: 0.9841
Epoch 8/20
1875/1875 [==============================] - 7s 4ms/step - loss: 0.0417 - accuracy: 0.9882
Epoch 9/20
1875/1875 [==============================] - 8s 4ms/step - loss: 0.0335 - accuracy: 0.9910
Epoch 10/20
1875/1875 [==============================] - 7s 4ms/step - loss: 0.027

In [34]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.10023392736911774
Test accuracy: 0.9782999753952026


In [35]:
# Using 16 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=16, epochs=20, verbose=1)


Epoch 1/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.4971 - accuracy: 0.8438
Epoch 2/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.1447 - accuracy: 0.9571
Epoch 3/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0944 - accuracy: 0.9722
Epoch 4/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0702 - accuracy: 0.9789
Epoch 5/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0517 - accuracy: 0.9841
Epoch 6/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0384 - accuracy: 0.9884
Epoch 7/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0284 - accuracy: 0.9916
Epoch 8/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0214 - accuracy: 0.9936
Epoch 9/20
3750/3750 [==============================] - 15s 4ms/step - loss: 0.0163 - accuracy: 0.9951
Epoch 10/20
3750/3750 [==============================] - 15s 4ms/step - l

In [36]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.09272510558366776
Test accuracy: 0.9814000129699707


In [37]:
# Using 8 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=8, epochs=20, verbose=1)


Epoch 1/20
7500/7500 [==============================] - 30s 4ms/step - loss: 0.3846 - accuracy: 0.8823
Epoch 2/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.1219 - accuracy: 0.9627
Epoch 3/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0787 - accuracy: 0.9759
Epoch 4/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0544 - accuracy: 0.9835
Epoch 5/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0419 - accuracy: 0.9872
Epoch 6/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0313 - accuracy: 0.9906
Epoch 7/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0237 - accuracy: 0.9928
Epoch 8/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0192 - accuracy: 0.9942
Epoch 9/20
7500/7500 [==============================] - 29s 4ms/step - loss: 0.0165 - accuracy: 0.9948
Epoch 10/20
7500/7500 [==============================] - 29s 4ms/step - l

In [38]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.09115327149629593
Test accuracy: 0.9837999939918518


The score increases as the batch size decreases, however the run time increases as well. It was more noticeable when the batch size got to 16 and lower. The best score was obtained from using a batch size of 8, but it took more than twice the time.

Choosing between batch sizes, I would recommend using a batch size of 64 because it provides an accuracy that is almost just as good as having a batch size of 8 but also executing in less than half the time it took for the 8 batch size model to run.



# ReLU Activation, SGD Loss, Batchsize of 64, and Changing the Amount of Neurons¶


In [39]:
# Having constant amount of 32 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(32, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 4s 4ms/step - loss: 1.5308 - accuracy: 0.4685
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.4928 - accuracy: 0.8496
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.3406 - accuracy: 0.8991
Epoch 4/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2810 - accuracy: 0.9174
Epoch 5/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2405 - accuracy: 0.9288
Epoch 6/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2151 - accuracy: 0.9352
Epoch 7/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1954 - accuracy: 0.9414
Epoch 8/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1780 - accuracy: 0.9471
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1658 - accuracy: 0.9512
Epoch 10/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1541 - accuracy: 0.9542

In [40]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.12300928682088852
Test accuracy: 0.9639999866485596


In [41]:
# Having constant amount of 64 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 4s 4ms/step - loss: 1.4956 - accuracy: 0.5310
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.4314 - accuracy: 0.8691
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.3122 - accuracy: 0.9076
Epoch 4/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2479 - accuracy: 0.9268
Epoch 5/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2075 - accuracy: 0.9390
Epoch 6/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1800 - accuracy: 0.9470
Epoch 7/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1595 - accuracy: 0.9525
Epoch 8/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1445 - accuracy: 0.9564
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1321 - accuracy: 0.9605
Epoch 10/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1217 - accuracy: 0.9632

In [42]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.11846434324979782
Test accuracy: 0.9625999927520752


In [43]:
# Having constant amount of 128 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 4s 4ms/step - loss: 1.2725 - accuracy: 0.6110
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.3524 - accuracy: 0.8946
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2575 - accuracy: 0.9239
Epoch 4/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2058 - accuracy: 0.9388
Epoch 5/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1727 - accuracy: 0.9485
Epoch 6/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1504 - accuracy: 0.9557
Epoch 7/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1326 - accuracy: 0.9601
Epoch 8/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1194 - accuracy: 0.9646
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1071 - accuracy: 0.9682
Epoch 10/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0992 - accuracy: 0.9704

In [44]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.10092046856880188
Test accuracy: 0.9710000157356262


In [45]:
# Having constant amount of 256 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(256, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 5s 5ms/step - loss: 1.2622 - accuracy: 0.6288
Epoch 2/20
938/938 [==============================] - 4s 5ms/step - loss: 0.3374 - accuracy: 0.9013
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2412 - accuracy: 0.9307
Epoch 4/20
938/938 [==============================] - 4s 5ms/step - loss: 0.1913 - accuracy: 0.9442
Epoch 5/20
938/938 [==============================] - 4s 5ms/step - loss: 0.1597 - accuracy: 0.9533
Epoch 6/20
938/938 [==============================] - 4s 5ms/step - loss: 0.1371 - accuracy: 0.9607
Epoch 7/20
938/938 [==============================] - 4s 5ms/step - loss: 0.1197 - accuracy: 0.9651
Epoch 8/20
938/938 [==============================] - 4s 5ms/step - loss: 0.1048 - accuracy: 0.9696
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0937 - accuracy: 0.9720
Epoch 10/20
938/938 [==============================] - 4s 5ms/step - loss: 0.0838 - accuracy: 0.9758

In [46]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.08184363692998886
Test accuracy: 0.9761999845504761


In [47]:
# Having constant amount of 512 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 6s 6ms/step - loss: 1.0332 - accuracy: 0.6894
Epoch 2/20
938/938 [==============================] - 5s 5ms/step - loss: 0.2944 - accuracy: 0.9131
Epoch 3/20
938/938 [==============================] - 5s 5ms/step - loss: 0.2177 - accuracy: 0.9362
Epoch 4/20
938/938 [==============================] - 5s 5ms/step - loss: 0.1735 - accuracy: 0.9492
Epoch 5/20
938/938 [==============================] - 5s 5ms/step - loss: 0.1445 - accuracy: 0.9579
Epoch 6/20
938/938 [==============================] - 5s 5ms/step - loss: 0.1229 - accuracy: 0.9642
Epoch 7/20
938/938 [==============================] - 5s 5ms/step - loss: 0.1056 - accuracy: 0.9689
Epoch 8/20
938/938 [==============================] - 5s 5ms/step - loss: 0.0925 - accuracy: 0.9725
Epoch 9/20
938/938 [==============================] - 5s 5ms/step - loss: 0.0809 - accuracy: 0.9762
Epoch 10/20
938/938 [==============================] - 5s 6ms/step - loss: 0.0718 - accuracy: 0.9783

In [48]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.0779823511838913
Test accuracy: 0.9754999876022339


In [49]:
# Having neurons decrease throughout model
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)


Epoch 1/20
938/938 [==============================] - 5s 4ms/step - loss: 1.2320 - accuracy: 0.6041
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.3578 - accuracy: 0.8938
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2317 - accuracy: 0.9323
Epoch 4/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1739 - accuracy: 0.9496
Epoch 5/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1401 - accuracy: 0.9598
Epoch 6/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1168 - accuracy: 0.9660
Epoch 7/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1013 - accuracy: 0.9701
Epoch 8/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0862 - accuracy: 0.9749
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0751 - accuracy: 0.9784
Epoch 10/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0661 - accuracy: 0.9808

In [50]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])


Test score: 0.09678570926189423
Test accuracy: 0.9740999937057495


When using constant amounts of neurons the test accuracy increases as the neurons increase, however, once the neurons get too high the run time starts to take longer to execute. The models all appear to have similar accuracies, but the model with a constant amound of 256 neurons performs the best in the most optimal run time. It even has a better accuracy than the model with neurons that decrease within each layer.

# Conclusion
Overall, each model is able to reach a very high accuracy, usually over 0.95. models that achieve an accuracy higher than 0.97 start to take a lot longer to execute, and only have a small increase in accuracy. Additionally, whenever the layers were increased the accuracy would increase, however once the layers got past 7 the accuracy would barely increase and the run time would take much longer. Therefore, I figured the most optimal amount of layers would be 7.

It appears the best activation function was the ReLU activation. I was able to obtain an accuracy score of around 0.977 with a quick run time using a constant amound of 256 neurons, ReLU activation, 7 layers, and SGD loss function.

# Possible Next Steps¶
Some next steps that could be done for improving the models would be to change the loss functions (incorporate a hinge loss function instead of categorical). Additionally, I could change the optimizer to something other than sgd (adam).